File: src/main/java/ltdjms/discord/redemption/domain/RedemptionCode.java
package ltdjms.discord.redemption.domain;

import java.time.Instant;
import java.util.Objects;

/**
 * Represents a redemption code that can be used to redeem a product. Each code is unique and can
 * only be used once.
 */
public record RedemptionCode(
    Long id,
    String code,
    Long productId,
    long guildId,
    Instant expiresAt,
    Long redeemedBy,
    Instant redeemedAt,
    Instant createdAt,
    Instant invalidatedAt,
    int quantity) {
  /** Length of the generated redemption code. */
  public static final int CODE_LENGTH = 16;

  /** Characters used for generating redemption codes. Excludes confusing characters: 0/O, 1/I/L */
  public static final String CODE_CHARACTERS = "ABCDEFGHJKMNPQRSTUVWXYZ23456789";

  public RedemptionCode {
    Objects.requireNonNull(code, "code must not be null");
    if (code.isBlank()) {
      throw new IllegalArgumentException("code must not be blank");
    }
    if (code.length() > 32) {
      throw new IllegalArgumentException("code must not exceed 32 characters");
    }
    // Ensure redeemed_by and redeemed_at are consistent
    if ((redeemedBy == null) != (redeemedAt == null)) {
      throw new IllegalArgumentException(
          "redeemedBy and redeemedAt must both be specified or both be null");
    }
    // Validate quantity
    if (quantity <= 0) {
      throw new IllegalArgumentException("quantity must be positive");
    }
    if (quantity > 1000) {
      throw new IllegalArgumentException("quantity must not exceed 1000");
    }
  }

  /**
   * Creates a new redemption code for a product. The ID will be null until the code is persisted.
   *
   * @param code the redemption code string
   * @param productId the ID of the product this code redeems
   * @param guildId the Discord guild ID
   * @param expiresAt the expiration time (can be null for no expiration)
   * @return a new RedemptionCode instance
   */
  public static RedemptionCode create(
      String code, long productId, long guildId, Instant expiresAt) {
    return create(code, productId, guildId, expiresAt, 1);
  }

  /**
   * Creates a new redemption code for a product with specified quantity. The ID will be null until
   * the code is persisted.
   *
   * @param code the redemption code string
   * @param productId the ID of the product this code redeems
   * @param guildId the Discord guild ID
   * @param expiresAt the expiration time (can be null for no expiration)
   * @param quantity the quantity of products this code redeems
   * @return a new RedemptionCode instance
   */
  public static RedemptionCode create(
      String code, long productId, long guildId, Instant expiresAt, int quantity) {
    return new RedemptionCode(
        null,
        code.toUpperCase(),
        productId,
        guildId,
        expiresAt,
        null,
        null,
        Instant.now(),
        null,
        quantity);
  }

  /**
   * Creates a copy of this code marked as redeemed.
   *
   * @param userId the user ID who redeemed the code
   * @return a new RedemptionCode instance marked as redeemed
   */
  public RedemptionCode withRedeemed(long userId) {
    if (isRedeemed()) {
      throw new IllegalStateException("Code has already been redeemed");
    }
    return new RedemptionCode(
        this.id,
        this.code,
        this.productId,
        this.guildId,
        this.expiresAt,
        userId,
        Instant.now(),
        this.createdAt,
        this.invalidatedAt,
        this.quantity);
  }

  /**
   * Checks if this code has been redeemed.
   *
   * @return true if the code has been used
   */
  public boolean isRedeemed() {
    return redeemedBy != null;
  }

  /**
   * Checks if this code has expired.
   *
   * @return true if the code has an expiration date and it has passed
   */
  public boolean isExpired() {
    return expiresAt != null && Instant.now().isAfter(expiresAt);
  }

  /**
   * Checks if this code is valid for use (not redeemed, not expired, and not invalidated).
   *
   * @return true if the code can be redeemed
   */
  public boolean isValid() {
    return !isInvalidated() && !isRedeemed() && !isExpired();
  }

  /**
   * Checks if this code has been invalidated. A code is invalidated when its associated product is
   * deleted.
   *
   * @return true if the code has been invalidated
   */
  public boolean isInvalidated() {
    return invalidatedAt != null;
  }

  /**
   * Creates a copy of this code marked as invalidated.
   *
   * @return a new RedemptionCode instance marked as invalidated
   */
  public RedemptionCode withInvalidated() {
    if (isInvalidated()) {
      throw new IllegalStateException("Code has already been invalidated");
    }
    return new RedemptionCode(
        this.id,
        this.code,
        null,
        this.guildId,
        this.expiresAt,
        this.redeemedBy,
        this.redeemedAt,
        this.createdAt,
        Instant.now(),
        this.quantity);
  }

  /**
   * Checks if this code belongs to the specified guild.
   *
   * @param checkGuildId the guild ID to check
   * @return true if this code belongs to the guild
   */
  public boolean belongsToGuild(long checkGuildId) {
    return this.guildId == checkGuildId;
  }

  /**
   * Gets a masked version of the code for display purposes. Shows only the first 4 and last 4
   * characters.
   *
   * @return masked code string
   */
  public String getMaskedCode() {
    if (code.length() <= 8) {
      return code;
    }
    return code.substring(0, 4) + "****" + code.substring(code.length() - 4);
  }
}

File: packages/shop/src/domain/redemption-code.ts
import { z } from 'zod';
import { ok, err, DomainError, type Result, safeSnowflakeToNumber } from '@ltdjms/shared';

/** Length of generated redemption codes. */
export const CODE_LENGTH = 16;

/** Characters for generating codes. Excludes confusing chars: 0/O, 1/I/L */
export const CODE_CHARACTERS = 'ABCDEFGHJKMNPQRSTUVWXYZ23456789';

/** Zod schema for RedemptionCode. */
export const RedemptionCodeSchema = z
  .object({
    id: z.number().int().positive().nullable(),
    code: z.string().min(1, 'code must not be blank').max(32, 'code must not exceed 32 characters'),
    productId: z.number().int().nullable(),
    guildId: z.number(),
    expiresAt: z.date().nullable(),
    redeemedBy: z.number().int().nullable(),
    redeemedAt: z.date().nullable(),
    createdAt: z.date(),
    invalidatedAt: z.date().nullable(),
    quantity: z.number().int().min(1, 'quantity must be positive').max(1000, 'quantity must not exceed 1000'),
  })
  .refine(
    (data) => (data.redeemedBy === null) === (data.redeemedAt === null),
    {
      message: 'redeemedBy and redeemedAt must both be specified or both be null',
    },
  );

export type RedemptionCode = z.infer<typeof RedemptionCodeSchema>;

export function createRedemptionCode(
  code: string,
  productId: number,
  guildId: number,
  expiresAt: Date | null,
  quantity = 1,
): RedemptionCode {
  const validated = RedemptionCodeSchema.parse({
    id: null,
    code: code.toUpperCase(),
    productId,
    guildId,
    expiresAt,
    redeemedBy: null,
    redeemedAt: null,
    createdAt: new Date(),
    invalidatedAt: null,
    quantity,
  });
  return validated;
}

export function withRedeemed(code: RedemptionCode, userId: string): Result<RedemptionCode, DomainError> {
  if (isRedeemed(code)) {
    return err(DomainError.invalidInput('Code has already been redeemed'));
  }
  return ok(RedemptionCodeSchema.parse({
    ...code,
    redeemedBy: safeSnowflakeToNumber(userId),
    redeemedAt: new Date(),
  }));
}

export function isRedeemed(code: RedemptionCode): boolean {
  return code.redeemedBy !== null;
}

export function isExpired(code: RedemptionCode): boolean {
  return code.expiresAt !== null && new Date() > code.expiresAt;
}

export function isValid(code: RedemptionCode): boolean {
  return !isInvalidated(code) && !isRedeemed(code) && !isExpired(code);
}

export function isInvalidated(code: RedemptionCode): boolean {
  return code.invalidatedAt !== null;
}

export function withInvalidated(code: RedemptionCode): Result<RedemptionCode, DomainError> {
  if (isInvalidated(code)) {
    return err(DomainError.invalidInput('Code has already been invalidated'));
  }
  return ok(RedemptionCodeSchema.parse({
    ...code,
    productId: null,
    invalidatedAt: new Date(),
  }));
}

export function belongsToGuild(code: RedemptionCode, guildId: number): boolean {
  return code.guildId === guildId;
}

export function getMaskedCode(code: RedemptionCode): string {
  if (code.code.length <= 8) return code.code;
  return code.code.substring(0, 4) + '****' + code.code.substring(code.code.length - 4);
}
